man/pmerge.1.rst
========
 pmerge
========

---------------------------
package merging, unmerging
---------------------------

:Date:   2007-02-26
:Manual section: 1
:Manual group: pkgcore

SYNOPSIS
========

  pmerge [options]

DESCRIPTION
===========

pmerge is the commandline interface for dependency resolution; merging, unmerging.
Currently under fairly heavy development, output/options will be improved in later
releases.

OPTIONS
=======

.. pkgcore_script_options:: pkgcore.scripts.pmerge.OptionParser


NEWS
Pkgcore Release Notes
=====================

See ChangeLog for full commit logs; this is summarized/major changes.

pkgcore 0.2.8:

* misc initialization fixes to syncers for when invoked via GenericSyncer.
  If you were getting __init__ tracebacks under layman integration, this
  is fixed.

* fallback bundled copy of elementtree import path was invalid, corrected.
  If you've been hitting any import issues with layman, this fixes it.

* shift the misc fs property triggers to pre_merge, rather then sanity_check;
  sanity_check should be only for "do I have what I need to even do the merge?".
  running preinst was occasionally wiping the changes the triggers made, thus
  allowing screwed up ebuilds with custom preinst's to slip in a portage gid
  for merge.

* fix a corner case for cpy join spotted by TFKyle where length calculation
  was incorrect, leading to a trailing null slipping into the calculated
  path.

* fix bash parsing for a corner case for empty assigns; literally, 
  x=
  foo='dar'
  would incorrectly interpret x=foo, instead of x=''.


pkgcore 0.2.7:

* layman configuration (if available) is now read for portage configuration
  for sync URI for overlays.  tar syncer is currently unsupported; others may
  be buggy.  Feed back desired (author doesn't use layman).  Ticket #11.  If
  you want it disabled, add FEATURES=-layman-sync .

* another fix for daft tarballs that try to touch cwd.

pkgcore 0.2.6:

* make intersecting ~ and =* atoms work again (used by pquery --revdep)

* catch a corner case py2.5 bug where AttributeError bleeds through from
  generic_equality.

* Via solars prodding, finished up the remaining bits for ROOT support.

* resolver traceback for if a requested atom is already known as insoluable.
  Thanks to kojiro for spotting it.

* misc bash code cleanup.

* PATH protection has been loosened slightly to enable 'weird' eclasses that
  are doing global PATH mangling.

* $HOME location for building was shifted into the targeted packages 
  directory, rather then a shared within $PORTAGE_TMPDIR.

* setgid/setuid triggers now match portage behaviour; -s,o-w mode change.

* trigger warnings are now enabled.

* New default trigger added; CommonDirectoryModes, checks for common
  directories (/usr, /etc, /usr/bin, /usr/lib for example) in the merge set,
  checking the packages specified modes for them.  If not 0755, throws a 
  warning.

* For directory on directory merging, ensure_perms (default op) was changed
  to preserve the existing directories permissions.  Generally speaking, this
  means that later versions of an ebuild have to use post_inst to correct the
  perms if they're incorrect- previously, the new perms/mode were forced on
  the existing.  Several common ebuilds (openssl for example) will generate
  weird modes on common directories however (heavily restricted perms), which
  can break things.  For the time being, the default is scaled down to the
  looser form portage does.

* added man page generation: pquery, pmerge

* pconfig now has a "dump-uncollapsed" command to dump the "raw" config.

* pebuild now supports --no-auto to run just the targeted phase.

* mass expansion of test coverage: pkgcore.restrictions.*,
  pkgcore.util.*, pkgcore.ebuild.*

* minor cleanup of pkgcore.test.ebuild.test_cpv to reduce redundant data sets;
  total testcase runtime reduction by about a third.

* diverge from unittest.TestCase to provide extra checks for normal asserts-
  assertNotEqual for example, checks both __eq__ and __ne__ now to smoke out
  any potential oversights in object equality implementation.

* use nsec mtime resolution if available to match python stdlib.

* env var PORTAGE_DEBUG for controlling how much debug info the ebuild env
  generates is now PKGCORE_DEBUG; range is the same, 0 (none), 1 (just the 
  ebuild/eclass), 2 (1 + relevant setup code), 3 (2 + filter-env data),
  4 (everything).


pkgcore 0.2.5:

* handle corner case in depend cycle processing where a package directly
  depends upon itself; fixes processing of sys-devel/libtool specifically.

* for pquery --attr keywords, sort by arch, not by stable/unstable.

* correct misc corner case atom bugs; an intersection bug, miss on an invalid 
  use dep atom lacking a closure in cpy atom, verification of use chars in
  native atom,

* osutils extensions tests, correcting a few cpy differences in behaviour from
  native.

* For unpacking a tarball that doesn't have it's files in a subdir, tar will
  occasionally try to utime the cwd resulting in a failure- uid owner for 
  WORKDIR was changed to allow tar to do the utime, thus succeed in unpacking.
  Only visible for userpriv and with oddball packages, gnuconfig for example.

* Cleanup of a few slow test cases; running the test suite should now be around
  25%-33% faster.

pkgcore/scripts/pmerge.py
# Copyright: 2006-2007 Brian Harring <ferringb@gmail.com>
# Copyright: 2006 Marien Zwart <marienz@gentoo.org>
# License: GPL2


"""Mess with the resolver and vdb."""


import time

from pkgcore.restrictions import packages, values
from pkgcore.util import commandline, parserestrict, lists, repo_utils
from pkgcore.util.compatibility import any
from pkgcore.ebuild import resolver
from pkgcore.repository import multiplex
from pkgcore.interfaces import observer, format
from pkgcore.util.formatters import ObserverFormatter
from pkgcore.util.packages import get_raw_pkg
from pkgcore.pkgsets.glsa import KeyedAndRestriction
from pkgcore.ebuild.atom import atom

class OptionParser(commandline.OptionParser):

    def __init__(self, **kwargs):
        commandline.OptionParser.__init__(self, description=__doc__, **kwargs)
        self.add_option('--deep', '-D', action='store_true',
            help='force the resolver to verify already installed dependencies')
        self.add_option('--unmerge', '-C', action='store_true',
            help='unmerge a package')
        self.add_option('--clean', action='store_true',
            help='remove installed packages that are not referenced by any '
            'target packages/sets; defaults to -s world -s system if no targets'
            ' are specified.  Use with *caution*, this option used incorrectly '
            'can render your system unusable.  Implies --deep'),
        self.add_option('--upgrade', '-u', action='store_true',
            help='try to upgrade already installed packages/depencies')
        self.add_option('--set', '-s', action='append',
            help='specify a pkgset to use')
        self.add_option('--ignore-failures', action='store_true',
            help='ignore resolution failures')
        self.add_option('--preload-vdb-state', action='store_true',
            help=\
"""enable preloading of the installed packages database
This causes the resolver to work with a complete graph, thus disallowing
actions that confict with installed packages.  If disabled, it's possible
for the requested action to conflict with already installed dependencies
that aren't involved in the graph of the requested operation""")

        self.add_option('--pretend', '-p', action='store_true',
            help="do the resolution, but don't merge/fetch anything")
        self.add_option('--ask', '-a', action='store_true',
            help="do the resolution, but ask to merge/fetch anything")
        self.add_option('--fetchonly', '-f', action='store_true',
            help="do only the fetch steps of the resolved plan")
        self.add_option('--ignore-cycles', '-i', action='store_true',
            help=\
"""ignore cycles if they're found to be unbreakable;
a depends on b, and b depends on a, with neither built is an example""")

        self.add_option('-B', '--with-built-depends', action='store_true',
            default=False,
            help="whether or not to process build depends for pkgs that "
            "are already built; defaults to ignoring them"),
        self.add_option('--nodeps', action='store_true',
            help='disable dependency resolution')
        self.add_option('--noreplace', action='store_false',
            dest='replace', default=True,
            help="don't reinstall target atoms if they're already installed")
        self.add_option('--usepkg', '-k', action='store_true',
            help="prefer to use binpkgs")
        self.add_option('--usepkgonly', '-K', action='store_true',
            help="use only built packages")
        self.add_option('--empty', '-e', action='store_true',
            help="force rebuilding of all involved packages, using installed "
                "packages only to satisfy building the replacements")
        self.add_option('--force', action='store_true',
                        dest='force',
            help="force merging to a repo, regardless of if it's frozen")
        self.add_option('--oneshot', '-o', action='store_true',
            default=False,
            help="do not record changes in the world file; if a set is "
            "involved, defaults to forcing oneshot")


    def check_values(self, options, args):
        options, args = commandline.OptionParser.check_values(
            self, options, args)
        options.targets = args

        if options.unmerge:
            if options.set:
                self.error("Sorry, using sets with -C probably isn't wise")
            if options.upgrade:
                self.error("can't combine upgrade and unmerging")
            if not options.targets:
                self.error("need at least one atom")
            if options.clean:
                self.error("Sorry, -C cannot be used with --clean")
        if options.clean:
            if options.set or options.targets:
                self.error("--clean currently has set/targets disabled; in "
                    "other words, accepts no args")
            options.set = ['world', 'system']
            options.deep = True
            if options.usepkgonly or options.usepkg:
                self.error(
                    '--usepkg and --usepkgonly cannot be used with --clean')
        elif options.usepkgonly and options.usepkg:
            self.error('--usepkg is redundant when --usepkgonly is used')
        if options.set:
            options.replace = False
        if not options.targets and not options.set:
            self.error('Need at least one atom/set')
        return options, ()


class AmbiguousQuery(parserestrict.ParseError):
    def __init__(self, token, keys):
        parserestrict.ParseError.__init__(
            self, '%s: multiple matches (%s)' % (token, ', '.join(keys)))
        self.token = token
        self.keys = keys

class NoMatches(parserestrict.ParseError):
    def __init__(self, token):
        parserestrict.ParseError.__init__(self, '%s: no matches' % (token,))

def parse_atom(token, repo, return_none=False):
    """Use L{parserestrict.parse_match} to produce a single atom.

    This matches the restriction against the repo, raises
    AmbiguousQuery if they belong to multiple cat/pkgs, returns an
    atom otherwise.

    @param token: string to convert.
    @param repo: L{pkgcore.repository.prototype.tree} instance to search in.
    @param return_none: indicates if no matches raises or returns C{None}

    @return: an atom or C{None}.
    """
    # XXX this should be in parserestrict in some form, perhaps.
    restriction = parserestrict.parse_match(token)
    key_matches = set(x.key for x in repo.itermatch(restriction))
    if not key_matches:
        raise NoMatches(token)
    elif len(key_matches) > 1:
        raise AmbiguousQuery(token, sorted(key_matches))
    if isinstance(restriction, atom):
        # atom is guranteed to be fine, since it's cat/pkg
        return restriction
    return KeyedAndRestriction(restriction, key=key_matches.pop())


class Failure(ValueError):
    """Raised internally to indicate an "expected" failure condition."""


def userquery(prompt, out, err, responses=None, default_answer=None, limit=3):
    """Ask the user to choose from a set of options.

    Displays a prompt and a set of responses, then waits for a
    response which is checked against the responses. If there is an
    unambiguous match the value is returned.

    @type prompt: C{basestring}.
    @type out: formatter.
    @type err: file-like object.
    @type responses: mapping with C{basestring} keys
    @param responses: mapping of user input to function result.
        Defaults to {"Yes": True, "No": False}.
    @param default_answer: returned if there is no input
        (user just hits enter). Defaults to True if responses is unset,
        unused otherwise.
    @param limit: number of allowed tries.
    """
    if responses is None:
        responses = {'Yes': True, 'No': False}
    if default_answer is None:
        default_answer = True
    for i in range(limit):
        response = raw_input('%s [%s] ' % (prompt, '/'.join(responses)))
        if not response and default_answer is not None:
            return default_answer

        results = set(
            (key, value) for key, value in responses.iteritems()
            if key[:len(response)].upper() == response.upper())
        if not results:
            out.write('Sorry, response "%s" not understood.' % (response,))
        elif len(results) > 1:
            out.write('Response "%s" is ambiguous (%s)' % (
                    response, ', '.join(key for key, val in results)))
        else:
            return list(results)[0][1]

    raise Failure('You have input a wrong response too many times.')


def unmerge(out, err, vdb, tokens, options, world_set=None):
    """Unmerge tokens. hackish, should be rolled back into the resolver"""
    all_matches = set()
    for token in tokens:
        # Catch restrictions matching across more than one category.
        # Multiple matches in the same category are acceptable.

        # The point is that matching across more than one category is
        # nearly always unintentional ("pmerge -C spork" without
        # realising there are sporks in more than one category), but
        # matching more than one cat/pkg is impossible without
        # explicit wildcards.
        restriction = parserestrict.parse_match(token)
        matches = vdb.match(restriction)
        if not matches:
            raise Failure('Nothing matches %s' % (token,))
        categories = set(pkg.category for pkg in matches)
        if len(categories) > 1:
            raise parserestrict.ParseError(
                '%s matches in multiple categories (%s)' % (
                    token, ', '.join(set(pkg.key for pkg in matches))))
        all_matches.update(matches)

    matches = sorted(all_matches)
    out.write(out.bold, 'Unmerge:')
    out.prefix = [out.bold, ' * ', out.reset]
    for match in matches:
        out.write(match.cpvstr)
    out.prefix = []

    repo_obs = observer.file_repo_observer(ObserverFormatter(out))

    if options.pretend:
        return

    if (options.ask and not
        userquery("Would you like to unmerge these packages?", out, err)):
        return
    return do_unmerge(options, out, err, vdb, matches, world_set, repo_obs)

def do_unmerge(options, out, err, vdb, matches, world_set, repo_obs):
    if vdb.frozen:
        if options.force:
            out.write(
                out.fg(out.red), out.bold,
                'warning: vdb is frozen, overriding')
            vdb.frozen = False
        else:
            raise Failure('vdb is frozen')

    for idx, match in enumerate(matches):
        out.write("removing %i of %i: %s" % (idx + 1, len(matches), match))
        out.title("%i/%i: %s" % (idx + 1, len(matches), match))
        op = vdb.uninstall(match, observer=repo_obs)
        ret = op.finish()
        if not ret:
            if not options.ignore_failures:
                raise Failure('failed unmerging %s' % (match,))
            out.write(out.fg(out.red), 'failed unmerging ', match)
        update_worldset(world_set, match, remove=True)
    out.write("finished; removed %i packages" % len(matches))


def get_pkgset(config, err, setname):
    try:
        return config.pkgset[setname]
    except KeyError:
        err.write('No set called %r!\nknown sets: %r' %
            (setname, config.pkgset.keys()))
        return None

def update_worldset(world_set, pkg, remove=False):
    if world_set is None:
        return
    if remove:
        try:
            world_set.remove(pkg)
        except KeyError:
            # nothing to remove, thus skip the flush
            return
    else:
        world_set.add(pkg)
    world_set.flush()

def main(options, out, err):
    config = options.config
    if options.debug:
        resolver.plan.limiters.add(None)

    domain = config.get_default('domain')
    vdb = domain.all_vdbs

    # This mode does not care about sets and packages so bypass all that.
    if options.unmerge:
        world_set = None
        if not options.oneshot:
            world_set = get_pkgset(config, err, "world")
            if world_set is None:
                err.write("disable world updating via --oneshot, or fix your "
                    "config")
                return 1
        try:
            unmerge(
                out, err, vdb, options.targets, options, world_set)
        except (parserestrict.ParseError, Failure), e:
            out.error(str(e))
            return 1
        return

    all_repos = domain.all_repos
    repos = list(all_repos.trees)
    if options.usepkgonly or options.usepkg:
        if options.usepkgonly:
            repos = [
                repo for repo in all_repos.trees
                if getattr(repo, 'format_magic', None) != 'ebuild_src']
        else:
            repos = [
                repo for repo in all_repos.trees
                if getattr(repo, 'format_magic', None) == 'ebuild_built'] + [
                repo for repo in all_repos.trees
                if getattr(repo, 'format_magic', None) != 'ebuild_built']
        all_repos = multiplex.tree(*repos)

    atoms = []
    for setname in options.set:
        pkgset = get_pkgset(config, err, setname)
        if pkgset is None:
            return 1
        l = list(pkgset)
        if not l:
            out.write("skipping set %s: set is empty, nothing to update" % setname)
        else:
            atoms.extend(l)

    for token in options.targets:
        try:
            a = parse_atom(token, all_repos, return_none=True)
        except parserestrict.ParseError, e:
            out.error(str(e))
            return 1
        if a is None:
            if token in config.pkgset:
                out.error(
                    'No package matches for %r, but there is a set with '
                    'that name. Use -s to specify a set.' % (token,))
                return 2
            elif not options.ignore_failures:
                out.error('No matches for %r; ignoring' % token)
            else:
                return -1
        else:
            atoms.append(a)

    if not atoms:
        out.error('No targets specified- nothing to do')
        return 1

    atoms = lists.stable_unique(atoms)

    world_set = None
    if (not options.set or options.clean) and not options.oneshot:
        world_set = get_pkgset(config, err, 'world')
        if world_set is None:
            err.write("disable world updating via --oneshot, or fix your "
                "config")
            return 1

    if options.upgrade:
        resolver_kls = resolver.upgrade_resolver
    else:
        resolver_kls = resolver.min_install_resolver

    extra_kwargs = {}
    if options.empty:
        extra_kwargs['resolver_cls'] = resolver.empty_tree_merge_plan

    resolver_inst = resolver_kls(
        vdb, repos, verify_vdb=options.deep, nodeps=options.nodeps,
        drop_cycles=options.ignore_cycles, force_replacement=options.replace,
        process_built_depends=options.with_built_depends,
        **extra_kwargs)

    if options.preload_vdb_state:
        out.write(out.bold, ' * ', out.reset, 'Preloading vdb... ')
        vdb_time = time.time()
        resolver_inst.load_vdb_state()
        vdb_time = time.time() - vdb_time
    else:
        vdb_time = 0.0

    failures = []
    resolve_time = time.time()
    out.write(out.bold, ' * ', out.reset, 'Resolving...')
    out.title('Resolving...')
    for restrict in atoms:
#        print "\ncalling resolve for %s..." % restrict
        ret = resolver_inst.add_atom(restrict)
        if ret:
            out.error('Resolver returned %r' % (ret,))
            out.error('resolution failed')
            failures.append(restrict)
            if not options.ignore_failures:
                break
    resolve_time = time.time() - resolve_time
    if failures:
        out.write()
        out.write('Failures encountered:')
        for restrict in failures:
            out.error("failed '%s'" % (restrict,))
            out.write('potentials:')
            match_count = 0
            for r in repo_utils.get_raw_repos(repos):
                l = r.match(restrict)
                if l:
                    out.write(
                        "repo %s: [ %s ]" % (r, ", ".join(str(x) for x in l)))
                    match_count += len(l)
            if not match_count:
                out.write("no matches found in %s" % (repos,))
            out.write()
            if not options.ignore_failures:
                return 1

    if options.clean:
        out.write(out.bold, ' * ', out.reset, 'packages to remove')
        vset = set(vdb)
        len_vset = len(vset)
        vset.difference_update(y.pkg for y in
            resolver_inst.state.iter_ops(True))
        wipes = sorted(x for x in vset if x.package_is_real)
        for x in wipes:
            out.write("remove %s" % x)
        out.write()
        if len(wipes):
            out.write("removing %i packages of %i installed, %0.2f%%." %
                (len(wipes), len_vset, 100*(len(wipes)/float(len_vset))))
        else:
            out.write("no packages to remove")
        if options.pretend:
            return 0;
        if options.ask:
            if not userquery("do you wish to proceed (default answer is no)?",
                out, err, default_answer=False):
                return 1
            out.write()
        repo_obs = observer.file_repo_observer(ObserverFormatter(out))
        do_unmerge(options, out, err, vdb, wipes, world_set, repo_obs)
        return 0;

    out.write(out.bold, ' * ', out.reset, 'buildplan')
    changes = list(x for x in resolver_inst.state.iter_ops()
        if x.pkg.package_is_real)
    ops_count = {}
    for op in changes:
        ops_count.setdefault(op.desc, 0)
        ops_count[op.desc] += 1
        if op.desc == "replace":
            out.write("replace %s, %s" %
                (get_raw_pkg(op.old_pkg), get_raw_pkg(op.pkg)))
        else:
            out.write("%s %s" % (op.desc.ljust(7), get_raw_pkg(op.pkg)))

    out.write()
    out.write("%i ops: %s" % (sum(ops_count.itervalues()),
        ", ".join("%i %ss" % (ops_count[k], k) for k in sorted(ops_count))))
    out.write()
    out.title('Resolved')
    out.write(out.bold, '%.2f' % (resolve_time,), out.reset,
              ' seconds resolving')
    if vdb_time:
        out.write(out.bold, '%.2f' % (vdb_time,), out.reset,
                  ' seconds preloading vdb state')
    if options.pretend:
        return

    if (options.ask and not
        userquery("Would you like to merge these packages?", out, err)):
        return

    build_obs = observer.file_build_observer(ObserverFormatter(out))
    repo_obs = observer.file_repo_observer(ObserverFormatter(out))

    change_count = len(changes)
    for count, op in enumerate(changes):
        status_str = "%i/%i, %s" % (count + 1, change_count,
            get_raw_pkg(op.pkg))
        out.write("processing %i of %i: %s" % (count + 1, change_count,
            get_raw_pkg(op.pkg)))
        out.title("%i/%i: %s" % (count + 1, change_count, get_raw_pkg(op.pkg)))
        if op.desc != "remove":
            if not options.fetchonly:
                out.write("forcing cleaning of workdir")
            buildop = op.pkg.build(observer=build_obs, clean=True)
            if options.fetchonly:
                out.write("\n%i files required-" % len(op.pkg.fetchables))
                try:
                    ret = buildop.fetch()
                except (SystemExit, KeyboardInterrupt):
                    raise
                except Exception, e:
                    ret = e
                if ret != True:
                    out.error("got %s for a phase execution for %s" % (ret, op.pkg))
                    if not options.ignore_failures:
                        return 1
                buildop.cleanup()
                del buildop, ret
                continue

            out.write("building...")
            ret = None
            try:
                built_pkg = buildop.finalize()
                if built_pkg is False:
                    ret = built_pkg
            except format.errors, e:
                ret = e
            if ret is None:
                out.write()
                if op.desc == "replace":
                    out.write("replace:  %s with %s" % 
                        (get_raw_pkg(op.old_pkg), get_raw_pkg(built_pkg)))
                    i = vdb.replace(op.old_pkg, built_pkg, observer=repo_obs)
                else:
                    out.write("install: %s" % get_raw_pkg(built_pkg))
                    i = vdb.install(built_pkg, observer=repo_obs)
            else:
                out.error("failure building %s: %s" % (get_raw_pkg(op.pkg),
                    ret))
                if not options.ignore_failures:
                    return 1
                continue
            # force this explicitly- can hold onto a helluva lot more
            # then we would like.
            del built_pkg
        else:
            out.write("remove:  %s" % get_raw_pkg(op.pkg))
            i = vdb.uninstall(op.pkg, observer=repo_objs)
        ret = i.finish()
        if ret != True:
            out.error("got %s for a phase execution for %s" % (ret, op.pkg))
            if not options.ignore_failures:
                return 1
        buildop.cleanup()
        if world_set:
            if op.desc == "remove":
                update_worldset(world_set, op.pkg, remove=True)
            elif any(x.match(op.pkg) for x in atoms):
                update_worldset(world_set, op.pkg)
    out.write("finished")
    return 0
